domain/externalcontroller/state/state.go

```diff
@@ -127,57 +128,218 @@ func (st *State) UpdateExternalController(
+			return errors.Trace(err)
 		}
+	}
 
-		// TODO (manadart 2023-05-13): Check current implementation and see if
-		// we need to delete models as we do for addresses, or whether this
-		// (additive) approach is what we have now.
-
-		for _, modelUUID := range modelUUIDs {
-			q := `
+	// TODO (manadart 2023-05-13): Check current implementation and see if
+	// we need to delete models as we do for addresses, or whether this
+	// (additive) approach is what we have now.
+	for _, modelUUID := range modelUUIDs {
+		q := `
 INSERT INTO external_model (uuid, controller_uuid)
 VALUES (?, ?)
   ON CONFLICT(uuid) DO UPDATE SET controller_uuid=excluded.controller_uuid`[1:]
 
-			if _, err := tx.ExecContext(ctx, q, modelUUID, cID); err != nil {
+		if _, err := tx.ExecContext(ctx, q, modelUUID, cID); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
+func (st *State) ModelsForController(
+	ctx context.Context,
+	controllerUUID string,
+) ([]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	q := `
+SELECT 	uuid 
+FROM   	external_model 
+WHERE  	controller_uuid = ?`
+
+	var modelUUIDs []string
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.QueryContext(ctx, q, controllerUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		for rows.Next() {
+			var modelUUID string
+			if err := rows.Scan(&modelUUID); err != nil {
+				_ = rows.Close()
 				return errors.Trace(err)
 			}
+			modelUUIDs = append(modelUUIDs, modelUUID)
 		}
 
 		return nil
 	})
 
-	return errors.Trace(err)
+	return modelUUIDs, err
+}
+
+func (st *State) ControllersForModels(ctx context.Context, modelUUIDs []string) ([]externalcontroller.MigrationControllerInfo, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	modelBinds, modelVals := database.SliceToPlaceholder(modelUUIDs)
+
+	q := fmt.Sprintf(`
+SELECT  ctrl.uuid,  
+	ctrl.alias,
+	ctrl.ca_cert,
+	model.uuid AS model,
+	addrs.address AS address
+FROM external_controller AS ctrl	
+JOIN 	external_model AS model
+ON ctrl.uuid = model.controller_uuid
+LEFT JOIN external_controller_address AS addrs
+ON ctrl.uuid = addrs.controller_uuid
+WHERE model.uuid IN (%s)`, modelBinds)
+
+	var resultControllers []externalcontroller.MigrationControllerInfo
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.QueryContext(ctx, q, modelVals...)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Prepare structs for unique models and addresses for each
+		// controller.
+		uniqueModelUUIDs := make(map[string]map[string]string)
+		uniqueAddresses := make(map[string]map[string]string)
+		uniqueControllers := make(map[string]externalcontroller.MigrationControllerInfo)
+		for rows.Next() {
+			var controller MigrationControllerInfo
+			if err := rows.Scan(&controller.ID, &controller.Alias, &controller.CACert, &controller.ModelUUID, &controller.Addr); err != nil {
```

Comment:
We have a row type for this already. Populate it like we do in `ControllerForModel`. You could even add a `ModelUUID` to it to account for the extra column you're selecting.

Reply:
As discussed, this is not possible for the moment due to sqlair limitations, I have added a TODO so we can do this later, see 00b4929935c3ca2cbe55f54ec0315a28ea71f19a

Reply:
After discussing with Alastair, we came up with a solution to use sqlair here, we still have the limitation of the `IN` statements though: https://github.com/canonical/sqlair/pull/76.

See 5cf8aedb07ae847baf05c0380091ff001e23d858